src/fromager/pyproject.py

```diff
@@ -0,0 +1,170 @@
+"""Tooling for pyproject.toml"""
+
+import ast
+import logging
+import pathlib
+import sys
+import typing
+
+import tomlkit
+from packaging.requirements import Requirement
+
+from . import context
+
+logger = logging.getLogger(__name__)
+
+TomlDict = dict[str, typing.Any]
+
+# section / key names
+BUILD_SYSTEM = "build-system"
+BUILD_BACKEND = "build-backend"
+BUILD_REQUIRES = "requires"
+
+
+class PyprojectFixer:
+    """Auto-fixer for pyproject.toml
+
+    - add missing pyproject.toml
+    - add or update [build-system] map
+    - detect additional build requirements by parsing "setup.py"
+    - remove unwanted build requirements such as cmake
+    """
+
+    # TODO: load these values from a config file
+    import_map: typing.ClassVar[dict[str, Requirement]] = {
+        "ninja": Requirement("ninja"),
+        "packaging": Requirement("packaging"),
+        "pybind11": Requirement("pybind11"),
+        "setuptools": Requirement("setuptools>=72.1.0"),
+        "torch": Requirement("torch"),
+        "wheels": Requirement("wheels"),
+    }
+
+    remove_build_requires: typing.ClassVar[list[Requirement]] = [
+        # use system cmake
+        Requirement("cmake"),
+    ]
+
+    # default to setuptools's build backend
+    default_build_backend = "setuptools.build_meta"
+
+    def __init__(
+        self, ctx: context.WorkContext, req: Requirement, sdist_root_dir: pathlib.Path
+    ) -> None:
+        self.ctx = ctx  # future use
+        self.req = req
+        self.root = sdist_root_dir
+        self.pyproject_toml = self.root / "pyproject.toml"
+        self.setup_py = self.root / "setup.py"
+
+    def update(self):
+        doc = self.load()
+        build_system = self.default_builds_system(doc)
+        self.update_build_requires(build_system)
+        self.save(doc)
```

Comment:
Does tomlkit preserve comments? Maybe we don't care if we roundtrip the file and lose info?

Reply:
Yes, tomlkit preserves comments and even layout of the file. Fromager's `pyproject.toml` does roundtrip perfectly.

```python
>>> import pathlib, tomlkit, toml
>>> raw = pathlib.Path("pyproject.toml").read_text()
>>> tomlkit.dumps(tomlkit.loads(raw)) == raw
>>> print(tomlkit.dumps(tomlkit.loads(raw)))
>>> print(toml.dumps(toml.loads(raw)))
```
tomlkit keeps
```toml
[tool.setuptools.dynamic]
optional-dependencies.build = { file = ["requirements-build.txt"] }
optional-dependencies.mypy = { file = ["requirements-mypy.txt"] }
optional-dependencies.test = { file = ["requirements-test.txt"] }
dependencies = { file = ["requirements.txt"] }
```

while toml changes the lines to
```toml
[tool.setuptools.dynamic.optional-dependencies.build]
file = [ "requirements-build.txt",]

[tool.setuptools.dynamic.optional-dependencies.mypy]
file = [ "requirements-mypy.txt",]

[tool.setuptools.dynamic.optional-dependencies.test]
file = [ "requirements-test.txt",]
```

Reply:
Nice, let's switch!

Reply:
Let me make a separate PR for that.